docs/developer-guides/Scripting_Terraform.md

```diff
@@ -0,0 +1,176 @@
+- Offers code completion and command signature assistance via Make for enhanced CLI usability
+- Supports named arguments with default values for an improved coding experience
+- Allows the working directory to be controlled by either arguments or a predefined constant for flexible stack management
+- Features a command wrapper to improve the onboarding experience and ensure environmental consistency
+- Incorporates both a Git hook and a GitHub action to enforce code quality standards
+- Comes with the CI/CD pipeline workflow integration
+- Includes a file cleanup routine to efficiently remove temporary resources
+- Incorporates a ready-to-run example to demonstrate the module's capabilities
+- Integrates a code linting routine to ensure scripts are free from unintended side effects
+- Includes a verbose mode for in-depth troubleshooting and debugging
+- Incorporates a best practice guide
+
+## Key files
+
+- Scripts
+  - [terraform.lib.sh](../../scripts/terraform/terraform.lib.sh) A library code loaded by custom make targets and CLI scripts
+  - [terraform.mk](../../scripts/terraform/terraform.mk): Customised implementation of the Terraform routines loaded by the `scripts/init.mk` file
+  - [terraform.sh](../../scripts/terraform/terraform.sh): Terraform command wrapper
+- Configuration
+  - [.tool-versions](../../.tool-versions): Stores Terraform version to be used
+- Code quality gates
+  - [lint-terraform/action.yaml](../../.github/actions/lint-terraform/action.yaml): GitHub action
+  - [check-terraform-format.sh](../../scripts/githooks/check-terraform-format.sh): Git hook
+- Usage example
+  - Declarative infrastructure definition example [terraform-state-aws-s3](../../scripts/terraform/examples/terraform-state-aws-s3) to store Terraform state
+  - A set of [make targets](https://github.com/nhs-england-tools/repository-template/blob/main/scripts/terraform/terraform.mk#L44) to run the example
+
+## Usage
+
+### Quick start
+
+Run the example:
+
+TODO: vscode integration ?
+
+```shell
+# AWS console access setup
+export AWS_ACCESS_KEY_ID="..."
+export AWS_SECRET_ACCESS_KEY="..."
+export AWS_SESSION_TOKEN="..."
+```
+
+```shell
+$ make terraform-example-provision-aws-infrastructure
+
+Initializing the backend..
+...
+Plan: 5 to add, 0 to change, 0 to destroy.
+Saved the plan to: terraform.tfplan
+To perform exactly these actions, run the following command to apply:
+    terraform apply "terraform.tfplan"
+...
+Apply complete! Resources: 5 added, 0 changed, 0 destroyed.
+
+$ make terraform-example-destroy-aws-infrastructure
+
+...
+Plan: 0 to add, 0 to change, 5 to destroy.
+...
+Apply complete! Resources: 0 added, 0 changed, 5 destroyed.
+```
+
+### Your stack implementation
+
+Always follow [best practices for using Terraform](https://cloud.google.com/docs/terraform/best-practices-for-terraform) while providing infrastructure as code (IaC) for your service.
+
+TODO: Extend the wording ?
```

Comment:
The thinking behind the directory structure may not be apparent without reading and understanding the appropriate section in the google best practice guide: https://cloud.google.com/docs/terraform/best-practices-for-terraform#subdirectories
In a nutshell, the idea is that the modules section contains the shared/common configuration for the service and the individual environment folders call those modules and provide their own variables.
I've seen various teams use various structures for various (good) reasons so we need to be sure the idea behind this suggested structure and the way to use it is understood.

Reply:
Yeah, I think it would be good to understand what some of that reasoning behind the variation is.  Have you got pointers to projects so we can start collating?

Reply:
I've added a paragraph to expand on that. Thanks both.